department-of-veterans-affairs/vagov-content
# Content deploy fast

Resolves https://github.com/department-of-veterans-affairs/vets.gov-team/issues/14855

This pull request implements the Content Deployment. It does this in three stages -

### 1. Rebuild Dev/Staging
- Rebuilds the latest commit of `vets-website`
- This does nothing special, or optimal. We know that the build job `testing/vets-website/master` will always check out the latest version of `vagov-content`, so this repo can piggyback off of that to rebuild the latest commit of `vets-website`.

### 2. Rebuild Production
This is a weird case. Dev/Staging should always have the latest commit, so we can rely on step 1 to update those environments. Prod reflects the latest release, however, so it's a special case. We have to rebuild the site with a local vets-website reset to the commit SHA of the latest release. The latest content will be checked out into a sibling directory. Then we can rebuild the site, upload to S3, and deploy.

- Steps:
    1. Checks out the content repo
    2. Checks out vets-website (aliases to vagov-apps)
    3. Resets the local vets-website to the tag of the latest release
    4. Rebuilds the local vets-website
    5. Uploads to S3 under the commit SHA of the tag
    6. Deploys

### 3. Deploy to production (or a spoofed environment)

## To Test
1. Define `isMaster` as -
```
isMaster = env.BRANCH_NAME == 'content-deploy-fast'
```

2. Manually set a release/SHA inside `dir(appCodeRepo) {`
```
releaseTag = 'vets-website/v0.1.383'
def imageTag = java.net.URLDecoder.decode(releaseTag).replaceAll("[^A-Za-z0-9\\-\\_]", "-")

releaseCommit = 'dab158f1e0bd5776963c4fa459a55c8237445f86'
dockerImage = docker.build("${appCodeRepo}:${imageTag}")
```
You should should choose a commit whose archive is harmless to overwrite (preferably for any environment). The release should be the latest release. This isn't necessary, but otherwise, the GitHub API caps you to a certain number of API hits per hour.

3. **Comment out the deploy step**
4. Download the archive of the commit SHA from S3 and confirm it was processed okay

Optionally, you can change the environment and build job to test even further (such as by depoying to dev.)
